Start refactoring wdio-browserstack to TypeScript

Fabianopb opened the pull request:
## Proposed changes

Package refactoring to TypeScript as part of https://github.com/webdriverio/webdriverio/issues/5828
Starting with `index`, `constants` and `utils`

## Types of changes

- [x] TypeScript refactoring

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added necessary documentation (if appropriate)

### Reviewers: @webdriverio/project-committers

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -0,0 +1,64 @@
+    [key: string]: { capabilities: BrowserCapabilities };
+};
+
+type BrowserCapabilities = GenericBrowserCaps & {
```
We should use `WebDriver.Capabilities` types here instead of defining something custom

Fabianopb replied:
Thanks for the tip! I gave it another round using `WebDriver.Capabilities` and `WebdriverIO.BrowserObject`, however two things I'm still not sure:

1. Not all keys from `BROWSER_DESCRIPTION` are within `Webdriver.Capabilities` so I had to included that in the local type.

2. `getBrowserCapabilities` receives a `string` `browserName` used to index `browser` and `caps` to get a `Webdriver.Capabilities` object. The way I found to solve it was through the union with 
```
{
    [key: string]: { capabilities: Webdriver.Capabilities };
}
```

What do you think?

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -1,10 +1,17 @@
+    [key: string]: { capabilities: WebDriver.Capabilities };
+};
+
+type Browser = WebdriverIO.BrowserObject & {
```
Just use `WebdriverIO.BrowserObject | WebdriverIO.MultibrowserObject`

Fabianopb replied:
Thanks, fixed!

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -1,10 +1,17 @@
-import { BROWSER_DESCRIPTION } from './constants'
+import { BROWSER_DESCRIPTION, BrowserDescription } from './constants'
 
+type Capabilities = WebDriver.Capabilities & BrowserDescription & {
```
same, use `WebDriver.Capabilities | WebdriverIO.MultiremoteBrowserCapabilities`

Fabianopb replied:
Thanks, fixed, but apparently this has to be intersection instead of union! :)

christian-bromann reviewed (commented):
Tests are missing. Let me know if I can help.

christian-bromann commented:
FYI: I closed #6015 as there seem to be no activity. So we can rewrite the whole package with this PR.

Fabianopb commented:
> FYI: I closed #6015 as there seem to be no activity. So we can rewrite the whole package with this PR.

Alright, so one PR to rule them all!

Fabianopb commented:
> Tests are missing. Let me know if I can help.

You mean converting the tests to TS, right?

christian-bromann commented:
> You mean converting the tests to TS, right?

Yes

Fabianopb commented:
Converted utils tests today and I'm unsure about one thing. This function has one argument only in the signature, but there's one test using a second arg, probably a mistake that wasn't caught before?

![image](https://user-images.githubusercontent.com/15076656/101280888-79bdac00-37d4-11eb-8c15-96510a07828d.png)

Fabianopb commented:
Strange new error breaking tests 🤔 
```
Error: packages/devtools/src/index.ts(2,10): error TS2616: 'UAParser' can only be imported by using 'import UAParser = require("../node_modules/@types/ua-parser-js")' or a default import.
```

christian-bromann commented:
> Strange new error breaking tests 🤔 Seems unrelated to this PR

@Fabianopb I just fixed in on master

christian-bromann commented:
> a mistake that wasn't caught before?

Correct, just remove the argument.

Fabianopb commented:
> I just fixed in on master

Thanks, I have rebased the branch

> Correct, just remove the argument.

Removed!

I'll continue refactoring the other files...

Fabianopb commented:
Started to give a try to `launcher.test.ts` and there were a few issues that we have to solve:

The launcher's `onComplete` doesn't accept an argument, but there are tests passing arguments:
![image](https://user-images.githubusercontent.com/15076656/101653958-a0841880-3a48-11eb-82df-c9b526273ca0.png)

Same for `addArgs`, this class method doesn't exist in `browserstack-local` but there are tests relying on mocking it:
![image](https://user-images.githubusercontent.com/15076656/101654123-d1fce400-3a48-11eb-95e1-06ddc1eda184.png)

How do you want to proceed? Do we remove those?

christian-bromann commented:
> The launcher's `onComplete` doesn't accept an argument, but there are tests passing argument

We can remove the arguments here.

> Same for `addArgs`, this class method doesn't exist in `browserstack-local` but there are tests relying on mocking it:

Do we use `addArgs` in our code somewhere? If so we should continue to test it and it seems some of the typings for BS local are wrong.

Fabianopb commented:
> Do we use `addArgs` in our code somewhere? If so we should continue to test it and it seems some of the typings for BS local are wrong.

It seems we don't use it anywhere, a global search shows results only in this test file!
![image](https://user-images.githubusercontent.com/15076656/101743417-62800680-3ad3-11eb-9e52-1784b0b977ac.png)

christian-bromann commented on packages/wdio-browserstack-service/src/launcher.ts:
```diff
@@ -6,13 +6,27 @@ import logger from '@wdio/logger'
+
+type Capabilities = (ExtendedCapabilities | ExtendedCapabilities[])
+
 export default class BrowserstackLauncherService {
```
Let's ensure we implement according to the service standard:
```suggestion
export default class BrowserstackLauncherService implements WebdriverIO.ServiceInstance {
```

christian-bromann commented on packages/wdio-browserstack-service/src/launcher.ts:
```diff
@@ -72,13 +86,13 @@ export default class BrowserstackLauncherService {
         return Promise.race([
             new Promise((resolve, reject) => {
-                this.browserstackLocal.stop((err) => {
+                this.browserstackLocal?.stop((err: any) => {
```
```suggestion
                this.browserstackLocal?.stop((err: Error) => {
```

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -21,10 +24,11 @@ export function getBrowserDescription(cap) {
+ * @param browserName browser name in case of multiremote
  */
-export function getBrowserCapabilities(browser, caps, browserName) {
+export function getBrowserCapabilities(browser: Browser, caps: Capabilities = {}, browserName: string = '') {
```
We should not default to empty string
```suggestion
export function getBrowserCapabilities(
    browser: Browser,
    caps: Capabilities = {},
    browserName?: string
) {
```

Fabianopb replied:
Done and fixed types

christian-bromann commented on packages/wdio-browserstack-service/tsconfig.json:
```diff
@@ -3,7 +3,8 @@
     "outDir": "./build",
-    "rootDir": "./src"
+    "rootDir": "./src",
+    "types": ["../webdriver", "../webdriverio", "./"]
```
```suggestion
    "types": ["webdriverio", "./"]
```

christian-bromann commented:
@Fabianopb let me know if this is ready for merge or if you plan more changes. Thanks for all the work so far 👏

Fabianopb commented:
> @Fabianopb let me know if this is ready for merge or if you plan more changes. Thanks for all the work so far 👏

Fixed according to suggestions! There's still `service.js` to be converted, wanna do it in a separate PR?

I have noticed that `utils.test.ts` have type errors because the `browser` mock lacks properties from the type for `Browser` we defined in `utils.ts`. Should we create a mock or do you have any other suggestion?

I noticed also this new error:
```
Error: packages/webdriverio/src/utils/index.ts(165,30): error TS2349: This expression is not callable.
```

christian-bromann commented:
> There's still `service.js` to be converted, wanna do it in a separate PR?

We can have the whole package as one PR.

> I have noticed that `utils.test.ts` have type errors because the `browser` mock lacks properties from the type for `Browser` we defined in `utils.ts`. Should we create a mock or do you have any other suggestion?

Can you point out which test?

Fabianopb commented:
> Can you point out which test?

Basically all tests in `utils.test.ts`, this one for example is missing lots of properties in the `browser` mock object passed to the function, which contains only `{ capabilities: { browser: 'browser' } }`

![image](https://user-images.githubusercontent.com/15076656/102014339-0da3f080-3d5e-11eb-90d8-cbed7831cf53.png)

christian-bromann commented:
@Fabianopb in tests we don't necessarily need to have total type safety. It is absolutely fine to pretend objects to be from a certain type to keep tests readable and simple (e.g. check out [those tests](https://github.com/webdriverio/webdriverio/blob/cb-rewrite-runner/packages/wdio-runner/tests/index.test.ts#L68-L75) I've just worked on).

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+type Config = WebdriverIO.Config & { cucumberOpts?: { strict: boolean } };
+
+type Suite = {
+    title: string;
+}
+
+type Test = {
+    fullName: string,
+    parent: string,
+    title: string
+};
+
+type Context = any;
+
+type Specs = any;
+
+type Feature = {
+    document: {
+        feature: {
+            name: string;
+        }
+    }
+};
+
+type Results = {
+    error: Error;
+    passed: boolean
+    exception: any;
+    status: string;
+};
+
+type Pickle = {
+    name: string;
+}
+
+type MultiRemoteAction = (sessionId: string, browserName?: string) => ReturnType<GotRequestFunction> | Promise<void>;
+
+type SessionResponse = {
+    // eslint-disable-next-line camelcase
+    automation_session: {
```
@christian-bromann I have created types here just to simulate what would be needed to solve all the errors in this file, however I'm not sure where they should come from, I'd need some help here! :)

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+    }
+};
+
+type Results = {
```
use `WebdriverIO.TestResult` instead

Fabianopb replied:
Same deal, `TestResult` doesn't include the properties we use here, perhaps the one we're looking for is: https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-cucumber-framework/cucumber-framework.d.ts#L15-L17

christian-bromann replied:
[this](https://github.com/webdriverio/webdriverio/blob/master/packages/webdriverio/webdriverio-core.d.ts#L257-L263) interface should be used

Fabianopb replied:
I think it needs some augmentation as it lacks `exception` and `status`, I'll add that!

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+
 export default class BrowserstackService {
-    constructor (options = {}, caps, config) {
+    config: Config;
+    sessionBaseUrl: string;
+    failReasons: string[];
+    scenariosThatRan: string[];
+    preferScenarioName: boolean;
+    strict: boolean;
+    failureStatuses: string[];
+    caps: Capabilities;
+    browser?: Browser;
+    fullTitle?: string;
```
Can we make them private starting with an `_`?

```suggestion
    private _config: Config;
    private _sessionBaseUrl: string;
    private _failReasons: string[];
    private _scenariosThatRan: string[];
    private _preferScenarioName: boolean;
    private _strict: boolean;
    private _failureStatuses: string[];
    private _caps: Capabilities;
    private _browser?: Browser;
    private _fullTitle?: string;
```

Fabianopb replied:
Done and fixed all references throughout the file!

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+    caps: Capabilities;
+    browser?: Browser;
+    fullTitle?: string;
+    constructor (options: BrowserstackConfig = {}, caps: Capabilities, config: Config) {
```
defining like this allows us to remove `this.config = config` in the constructor
```suggestion
    constructor (options: BrowserstackConfig = {}, caps: Capabilities, private _config: Config) {
```

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+    fullTitle?: string;
+    constructor (options: BrowserstackConfig = {}, caps: Capabilities, config: Config) {
         this.config = config
         this.sessionBaseUrl = 'https://api.browserstack.com/automate/sessions'
```
let's define defaults in the class head

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+    }
+}
+
 export default class BrowserstackService {
```
```suggestion
export default class BrowserstackService implements WebdriverIO.ServiceInstance {
```

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+
+type Config = WebdriverIO.Config & { cucumberOpts?: { strict: boolean } };
+
+type Suite = {
```
I had actually tried this, but got errors:
![image](https://user-images.githubusercontent.com/15076656/102265239-81d9c200-3f1f-11eb-864e-96a9b662b162.png)

The `Suite` type used here, seems to come from `jasmine-framework.d.ts`: https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-jasmine-framework/jasmine-framework.d.ts#L7-L11

christian-bromann replied:
both Mocha and Jasmine suite interfaces have `title` properties.

Fabianopb replied:
I've augmented the interfaces in `browserstack-service.d.ts`

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+    title: string;
+}
+
+type Test = {
```
Same deal here, the type that should be used seems to be the one coming from here: https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-jasmine-framework/jasmine-framework.d.ts#L13-L16

christian-bromann replied:
Let's use `WebdriverIO.Test` for now like in other services and clean this up later.

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -1,12 +1,72 @@
+
+type Specs = any;
+
+type Feature = {
```
Hummm... It seems we don't have a type called `Feature` in `WebdriverIO` 🤔

christian-bromann replied:
right, nevermind then

christian-bromann commented:
@Fabianopb anything I can do to push this over the line?

Fabianopb commented:
Sorry @christian-bromann, I ended stalling this as holidays came! I think some types in tests were still broken if you want to take a look, otherwise I can return to this after New Year!

Fabianopb commented:
Happy New Year @christian-bromann ! 🎉 
I'm back to biz, just finished converting the last test file, please take a careful look to see if we have all in place now!

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    private _caps: Capabilities;
+    private _browser?: Browser;
+    private _fullTitle?: string;
+    constructor (options: BrowserstackConfig = {}, caps: Capabilities, private _config: WebdriverIO.Config) {
+        // Cucumber specific
+        this._preferScenarioName = Boolean(options.preferScenarioName)
+        this._strict = Boolean(_config.cucumberOpts && _config.cucumberOpts.strict)
+        // See https://github.com/cucumber/cucumber-js/blob/master/src/runtime/index.ts#L136
+        this._strict && this._failureStatuses.push('pending')
+        this._caps = caps
+    }
```
some improvements here:
```suggestion
    constructor (private _options: BrowserstackConfig = {}, private _caps: Capabilities, private _config: WebdriverIO.Config) {
        if (this._strict) {
            this._failureStatuses.push('pending')
        }
    }
```

`_strict` is nowhere else used so we can get rid of it and by making the whole options object a class property we don't need to specifically define `_preferScenarioName`. The comment pointed to line of code that doesn't really indicate any information as to why this was implemented like it.

Fabianopb replied:
Removing these will break some tests:
```ts
this._preferScenarioName = Boolean(_options.preferScenarioName)
this._strict = Boolean(_config.cucumberOpts && _config.cucumberOpts.strict)
```

christian-bromann replied:
the tests should be updated then.

Fabianopb replied:
Did some changes, but I still have a question, follow up in this comment: https://github.com/webdriverio/webdriverio/pull/6172#discussion_r553558912

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    after(result: number) {
+        // For Cucumber: Checks scenarios that ran (i.e. not skipped) on the session
+        // Only 1 Scenario ran and option enabled => Redefine session name to Scenario's name
+        if (this._preferScenarioName && this._scenariosThatRan.length === 1){
```
based on the suggestion above:
```suggestion
        if (this._options._preferScenarioName && this._scenariosThatRan.length === 1){
```

Fabianopb replied:
Done! (I didn't commit this suggestion as "preferScenarioName" shouldn't be suffixed by underscore)

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    private _sessionBaseUrl: string = 'https://api.browserstack.com/automate/sessions';
+    private _failReasons: string[] = [];
+    private _scenariosThatRan: string[] = [];
+    private _preferScenarioName: boolean;
+    private _strict: boolean;
```
With that, these aren't needed anymore
```suggestion
```

Fabianopb replied:
Removed from here and from tests

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    private _preferScenarioName: boolean;
+    private _strict: boolean;
+    private _failureStatuses: string[] = ['failed', 'ambiguous', 'undefined', 'unknown'];
+    private _caps: Capabilities;
```
with that this gets defined in the constructor.

Fabianopb replied:
👍

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    async onReload(oldSessionId: string, newSessionId: string) {
+        const { _browser } = this
```
I would prefer to just use `this._browser`

Fabianopb replied:
👍

christian-bromann commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    async _printSessionURL() {
+        const { _browser } = this
```
same

Fabianopb replied:
👍

christian-bromann commented on packages/wdio-browserstack-service/src/util.ts:
```diff
@@ -0,0 +1,47 @@
+import { BROWSER_DESCRIPTION } from './constants'
+
+type Capabilities = WebDriver.Capabilities & WebdriverIO.MultiRemoteCapabilities;
+
+type Browser = WebdriverIO.BrowserObject & WebdriverIO.MultiRemoteBrowserObject;
```
we have defined these in `service.ts` already. I suggest to move those into a `types.ts` and import them here.

christian-bromann reviewed (commented):
Thank you so much for all your work 😊 I added some comments, mostly optimisations that are not much related to the rewrite.

Fabianopb commented on packages/wdio-browserstack-service/src/launcher.ts:
```diff
@@ -6,13 +6,27 @@ import logger from '@wdio/logger'
+}
+
+type ExtendedCapabilities = WebDriver.Capabilities & {
+    'browserstack.local'?: boolean;
```
Done!

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+import logger from '@wdio/logger'
+import got, { GotRequestFunction } from 'got'
+import WebDriver from 'webdriver'
```
Removed WebDriver import

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,240 @@
+        // eslint-disable-next-line camelcase
+        browser_url: string
+    }
+}
```
Moved!

Fabianopb commented on packages/wdio-browserstack-service/src/service.ts:
```diff
@@ -0,0 +1,207 @@
+const log = logger('@wdio/browserstack-service')
+
+export default class BrowserstackService implements WebdriverIO.ServiceInstance {
+    private _sessionBaseUrl: string = 'https://api.browserstack.com/automate/sessions';
+    private _failReasons: string[] = [];
+    private _scenariosThatRan: string[] = [];
+    private _failureStatuses: string[] = ['failed', 'ambiguous', 'undefined', 'unknown'];
+    private _browser?: Browser;
+    private _fullTitle?: string;
+    constructor (private _options: BrowserstackConfig = {}, private _caps: Capabilities, private _config: WebdriverIO.Config) {
+        // Cucumber specific
+        const strict = Boolean(_config.cucumberOpts && _config.cucumberOpts.strict)
+        // See https://github.com/cucumber/cucumber-js/blob/master/src/runtime/index.ts#L136
+        if (strict) {
+            this._failureStatuses.push('pending')
+        }
+    }
```
@christian-bromann after the last changes one question remains: is this how this strict check should still happen?

christian-bromann replied:
I think so, what concerns do you have?

Fabianopb replied:
None, I think I had the impression from this suggestion (https://github.com/webdriverio/webdriverio/pull/6172#discussion_r551799101) that this line under "Cucumber specific" comment was to be removed.
```ts
// Cucumber specific
const strict = Boolean(_config.cucumberOpts && _config.cucumberOpts.strict)
```

Fabianopb replied:
So, this should be good to go then? 🙂

christian-bromann reviewed (approved):
Yes, this looks good! Thank you so much 🙏

Fabianopb commented:
> Yes, this looks good! Thank you so much 🙏

Yay! That was a long stretch! Thanks for the reviews! 🎉

reidj32 commented:
I realize this has been merged, but after doing a package upgrade today I found an issue which might be related to this and wanted to confirm before opening a new issue. I have both @wdio/browserstack-service (6.12.0) and @wdio/cucumber-framework (6.11.1) installed and they are using TypeScript. After taking the most recent upgrades, there appears to be a type collision in the WebdriverIO module. Here is the error I receive when compiling:
![image](https://user-images.githubusercontent.com/12374462/105075793-da850900-5a4f-11eb-8661-fe4055b685f1.png)

After some digging, I found that the same interface extension occurs in both of these packages which is causing the issue. If I open `node_modules/@wdio/browserstack-service/browserstack-service.d.ts` and make the following modifications, I am able to compile without issue.
```typescript
declare module WebdriverIO {
    interface ServiceOption extends BrowserstackConfig {}

    // interface Config extends CucumberOptsConfig {

    // }

    interface Suite {
        title: string;
        fullName: string;
        file: string;
    }

    interface Test extends Suite {
        parent: string;
        passed: boolean;
    }

    interface TestResult {
        exception: string;
        status: string;
    }
}

interface BrowserstackConfig {
    /**
     * Set this to true to enable routing connections from Browserstack cloud through your computer.
     * You will also need to set `browserstack.local` to true in browser capabilities.
     */
    browserstackLocal?: boolean;
    /**
     * Cucumber only. Set this to true to enable updating the session name to the Scenario name if only
     * a single Scenario was ran. Useful when running in parallel
     * with [wdio-cucumber-parallel-execution](https://github.com/SimitTomar/wdio-cucumber-parallel-execution).
     */
    preferScenarioName?: boolean;
    /**
     * Set this to true to kill the browserstack process on complete, without waiting for the
     * browserstack stop callback to be called. This is experimental and should not be used by all.
     */
    forcedStop?: boolean;
    /**
     * Specified optional will be passed down to BrowserstackLocal. See this list for details:
     * https://stackoverflow.com/questions/39040108/import-class-in-definition-file-d-ts
     */
    opts?: Partial<import('browserstack-local').Options>
}

// interface CucumberOptsConfig {
//     cucumberOpts?: CucumberOpts
// }

// interface CucumberOpts {
//     /**
//      * Fail if there are any undefined or pending steps
//      * @default false
//      */
//     strict?: boolean
// }
```

I am curious why the browserstack service cares about the cucumber types?

christian-bromann commented:
> I am curious why the browserstack service cares about the cucumber types?

Because it handles reporting job status differently depending whether [cucumberOpts are set](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-browserstack-service/src/service.ts#L18) to strict. Let's report this or also PRs would be appreciated 🙏
